doc/sphinx/language/core/records.rst

```diff
@@ -1,248 +1,343 @@
 
-.. coqtop:: all
+         Record MyRecord := { myfield : nat } as VarName.
+         About myfield. (* observe the MyRecord variable is named "VarName" *)
 
-  Definition half' :=
-    {| sign := true;
-       Rat_bottom_cond := O_S 1;
-       Rat_irred_cond := one_two_irred |}.
+         (* make "VarName" implicit without having to rename the variable,
+            which would be necessary without the "as" clause *)
+         Arguments myfield {VarName}.   (* make "VarName" an implicit parameter *)
+         Check myfield.
+         Check (myfield (VarName:={| myfield := 0 |})).
 
-The following settings let you control the display format for types:
+   .. exn:: Records declared with the keyword Record or Structure cannot be recursive.
 
-.. flag:: Printing Records
+      The record name :token:`ident` appears in the type of its fields, but uses
+      the :cmd:`Record` command. Use  the :cmd:`Inductive` or
+      :cmd:`CoInductive` command instead.
 
-   When this :term:`flag` is on (this is the default),
-   use the record syntax (shown above) as the default display format.
+   .. exn:: @ident already exists
 
-You can override the display format for specified types by adding entries to these tables:
+      The fieldname :n:`@ident` is already defined as a global.
 
-.. table:: Printing Record @qualid
+   .. warn:: @ident__1 cannot be defined because the projection @ident__2 was not defined
 
-   This :term:`table` specifies a set of qualids which are displayed as records.  Use the
-   :cmd:`Add` and :cmd:`Remove` commands to update the set of qualids.
+      The type of the projection :token:`ident` depends on previous projections which
+      themselves could not be defined.
 
-.. table:: Printing Constructor @qualid
+   .. warn:: @ident cannot be defined.
 
-   This :term:`table` specifies a set of qualids which are displayed as constructors.  Use the
-   :cmd:`Add` and :cmd:`Remove` commands to update the set of qualids.
+      The projection cannot be defined.  This message is followed by an explanation
+      of why it's not possible, such as:
 
-This syntax can also be used for pattern matching.
+      #. The :term:`body` of :token:`ident` uses an incorrect elimination for
+         :token:`ident` (see :cmd:`Fixpoint` and :ref:`Destructors`).
 
-.. coqtop:: all
+   .. warn:: @ident__field cannot be defined because it is informative and @ident__record is not
 
-   Eval compute in (
-     match half with
-     | {| sign := true; top := n |} => n
-     | _ => 0
-     end).
+      The projection for the named field can't be defined.
+      For example, :n:`Record foo:Prop := { x:Type }` generates the message
+      "x cannot be defined ... and foo is not".  Proofs (objects of sort :n:`Prop`)
+      are supposed to be non-distinguishable.  If you have two inhabitants of
+      :n:`Type`, such as :n:`%{%| x := nat %|%}` and :n:`%{%| x := bool %|%}`, they are
+      distinguishable (i.e. informative) and are therefore prohibited.
```

Comment:
An example using simpler objects (i.e. not a type of types) would be to simply take `Record foo:Prop := { x:nat }`, then continuing e.g. as follows: ``Inhabitants of :n:`nat` are distinguishable (i.e. informative), so to preserve say :n:`%{%| x := 0 %|%}` and :n:`%{%| x := 1 %|%}` from being distinguished, the projection has to be prohibited.``

We could even take `bool` instead of `nat` which  would be the smallest counterexample to leaking information via projections.

Reply:
How about this?

![image](https://user-images.githubusercontent.com/1253341/154819565-89510f86-7c7e-4667-90dd-cfa6cd78c01f.png)

(OTOH, I don't quite see how a record could be a proof.  No need to explain now.)

Reply:
I think that's ok (records -> record?)

> (OTOH, I don't quite see how a record could be a proof. No need to explain now.)

A typical example of proposition which it makes sense to defined as a `Record` is `Record and (A B:Prop) := {proj1:A ; proj2:B}.` (even if it is not done like this in the library).